In the 20×20 grid below, four numbers along a diagonal line have been marked in red.


> 08 02 22 97 38 15 00 40 00 75 04 05 07 78 52 12 50 77 91 08

> 49 49 99 40 17 81 18 57 60 87 17 40 98 43 69 48 04 56 62 00

> 81 49 31 73 55 79 14 29 93 71 40 67 53 88 30 03 49 13 36 65

> 52 70 95 23 04 60 11 42 69 24 68 56 01 32 56 71 37 02 36 91 

> 22 31 16 71 51 67 63 89 41 92 36 54 22 40 40 28 66 33 13 80 

> 24 47 32 60 99 03 45 02 44 75 33 53 78 36 84 20 35 17 12 50

> 32 98 81 28 64 23 67 10 26 38 40 67 59 54 70 66 18 38 64 70

> 67 26 20 68 02 62 12 20 95 63 94 39 63 08 40 91 66 49 94 21

> 24 55 58 05 66 73 99 26 97 17 78 78 96 83 14 88 34 89 63 72

> 21 36 23 09 75 00 76 44 20 45 35 14 00 61 33 97 34 31 33 95

> 78 17 53 28 22 75 31 67 15 94 03 80 04 62 16 14 09 53 56 92

> 16 39 05 42 96 35 31 47 55 58 88 24 00 17 54 24 36 29 85 57

> 86 56 00 48 35 71 89 07 05 44 44 37 44 60 21 58 51 54 17 58

> 19 80 81 68 05 94 47 69 28 73 92 13 86 52 17 77 04 89 55 40

> 04 52 08 83 97 35 99 16 07 97 57 32 16 26 26 79 33 27 98 66

> 88 36 68 87 57 62 20 72 03 46 33 67 46 55 12 32 63 93 53 69

> 04 42 16 73 38 25 39 11 24 94 72 18 08 46 29 32 40 62 76 36

> 20 69 36 41 72 30 23 88 34 62 99 69 82 67 59 85 74 04 36 16

> 20 73 35 29 78 31 90 01 74 31 49 71 48 86 81 16 23 57 05 54

> 01 70 54 71 83 51 54 69 16 92 33 48 61 43 52 01 89 19 67 48


The product of these numbers is 26 × 63 × 78 × 14 = 1788696.

What is the greatest product of four adjacent numbers in the same direction (up, down, left, right, or diagonally) in the 20×20 grid?

In [1]:
grid = """
08 02 22 97 38 15 00 40 00 75 04 05 07 78 52 12 50 77 91 08
49 49 99 40 17 81 18 57 60 87 17 40 98 43 69 48 04 56 62 00
81 49 31 73 55 79 14 29 93 71 40 67 53 88 30 03 49 13 36 65
52 70 95 23 04 60 11 42 69 24 68 56 01 32 56 71 37 02 36 91
22 31 16 71 51 67 63 89 41 92 36 54 22 40 40 28 66 33 13 80
24 47 32 60 99 03 45 02 44 75 33 53 78 36 84 20 35 17 12 50
32 98 81 28 64 23 67 10 26 38 40 67 59 54 70 66 18 38 64 70
67 26 20 68 02 62 12 20 95 63 94 39 63 08 40 91 66 49 94 21
24 55 58 05 66 73 99 26 97 17 78 78 96 83 14 88 34 89 63 72
21 36 23 09 75 00 76 44 20 45 35 14 00 61 33 97 34 31 33 95
78 17 53 28 22 75 31 67 15 94 03 80 04 62 16 14 09 53 56 92
16 39 05 42 96 35 31 47 55 58 88 24 00 17 54 24 36 29 85 57
86 56 00 48 35 71 89 07 05 44 44 37 44 60 21 58 51 54 17 58
19 80 81 68 05 94 47 69 28 73 92 13 86 52 17 77 04 89 55 40
04 52 08 83 97 35 99 16 07 97 57 32 16 26 26 79 33 27 98 66
88 36 68 87 57 62 20 72 03 46 33 67 46 55 12 32 63 93 53 69
04 42 16 73 38 25 39 11 24 94 72 18 08 46 29 32 40 62 76 36
20 69 36 41 72 30 23 88 34 62 99 69 82 67 59 85 74 04 36 16
20 73 35 29 78 31 90 01 74 31 49 71 48 86 81 16 23 57 05 54
01 70 54 71 83 51 54 69 16 92 33 48 61 43 52 01 89 19 67 48
"""

grid = grid.strip().split("\n")
grid = [line.split(" ") for line in grid]

In [7]:
from functools import reduce
from operator import mul
# total, start coordinates, direction
largest = (0, None, None)
def product(cells):
    # given a list of strings representing cells, return product
    return reduce(mul, map(int, cells))
    
side = len(grid)
g = 4
if len(grid[0]) != side:
    raise ValueError("Bad format")
for r, row in enumerate(grid):
    for c, cell in enumerate(row):
        if c <= side-g:
            # calculate right neighbours. don't need left because
            # it will already have been calculated
            cells = row[c:c+g]
            current = product(cells)
            if current > largest[0]:
                largest = (current, (r, c), 'right')
        if r <= side-g:
            # calculate down neighbours
            cells = [grid[r+i][c] for i in range(0, g)]
            current = product(cells)
            if current > largest[0]:
                largest = (current, (r, c), 'down')
        if r <= side-g and c <= side-g:
            # calculate diagonal in down, right direction
            cells = [grid[r+i][c+i] for i in range(0, g)]
            current = product(cells)
            if current > largest[0]:
                largest = (current, (r, c), 'down/right')
        if r <= side-g and c >= g:
            # calculate diagonal in down, left direction
            cells = [grid[r+i][c-i] for i in range(0, g)]
            current = product(cells)
            if current > largest[0]:
                largest = (current, (r, c), 'down/left')

largest

(70600674, (12, 6), 'down/left')

The sequence of triangle numbers is generated by adding the natural numbers. So the 7th triangle number would be 1 + 2 + 3 + 4 + 5 + 6 + 7 = 28. The first ten terms would be:

1, 3, 6, 10, 15, 21, 28, 36, 45, 55, ...

Let us list the factors of the first seven triangle numbers:

 1: 1
 
 3: 1,3
 
 6: 1,2,3,6

10: 1,2,5,10

15: 1,3,5,15

21: 1,3,7,21

28: 1,2,4,7,14,28

We can see that 28 is the first triangle number to have over five divisors.

What is the value of the first triangle number to have over five hundred divisors?

In [54]:
def factorise(n):
    if n <= 0:
        raise ValueError("Only works for positive n: got {}".format(n))
    if n == 1:
        return n
    factors = []
    factor = 2
    # get rid of 2s
    # in separate loop because 
    # in main loop we have to increment by 2,
    # here only increment by 1. 
    while n %2 == 0:
        factors.append(2)
        n = n//2

    # try all odd numbers.
    factor = 3
    while factor <= n:
        #print(n)
        while n%factor == 0:
            factors.append(factor)
            n = n // factor
        factor += 2
    return factors

def powerset(seq):
    # use binary count. 
    n = len(seq)
    retval = []
    for i in range(0, 2**n):
        thisval = []
        # work out what binary maps to this number
        bitstring = '{:b}'.format(i).zfill(n)
        for idx, bit in enumerate(bitstring):
            if int(bit):
               thisval.append(seq[idx])
        if thisval:
            retval.append(thisval)
    return retval

def divisors(n):
    # find prime factors
    # take all set combinations of those prime factors
    pfactors = factorise(n)
    combinations = powerset(pfactors)
    #print(combinations)
    factors = [reduce(mul, x) for x in combinations]
    factors = [1] + factors + [n]
    return set(factors)

def triangle_number_stop(ndivs=5):
    # return the first triangle number with greater than ndivs divisors
    n = 1
    total = 1
    divs = {1}
    while len(divs) < ndivs:
        n += 1
        total += n
        divs = divisors(total)
        print(total, divs)
    return total
    
    

In [55]:
triangle_number_stop(500)

3 {1, 3}
6 {1, 2, 3, 6}
10 {1, 2, 10, 5}
15 {1, 3, 5, 15}
21 {1, 3, 21, 7}
28 {1, 2, 4, 7, 14, 28}
36 {1, 2, 3, 4, 36, 6, 9, 12, 18}
45 {1, 3, 5, 9, 45, 15}
55 {1, 11, 5, 55}
66 {1, 2, 3, 33, 66, 6, 11, 22}
78 {1, 2, 3, 6, 39, 13, 78, 26}
91 {1, 91, 13, 7}
105 {1, 3, 35, 5, 7, 105, 15, 21}
120 {1, 2, 3, 4, 5, 6, 8, 40, 10, 12, 120, 15, 20, 24, 60, 30}
136 {1, 2, 34, 4, 68, 8, 136, 17}
153 {1, 3, 9, 17, 51, 153}
171 {1, 3, 9, 171, 19, 57}
190 {1, 2, 5, 38, 10, 19, 190, 95}
210 {1, 2, 3, 35, 5, 70, 7, 6, 105, 10, 42, 14, 15, 210, 21, 30}
231 {1, 33, 3, 7, 231, 11, 77, 21}
253 {1, 11, 253, 23}
276 {1, 2, 3, 4, 69, 6, 138, 12, 46, 276, 23, 92}
300 {1, 2, 3, 4, 5, 6, 100, 10, 75, 12, 300, 15, 50, 20, 150, 25, 60, 30}
325 {65, 1, 5, 325, 13, 25}
351 {1, 3, 39, 9, 13, 117, 27, 351}
378 {1, 2, 3, 6, 7, 9, 42, 14, 18, 21, 54, 378, 27, 189, 126, 63}
406 {1, 2, 7, 203, 14, 406, 58, 29}
435 {1, 3, 5, 15, 145, 435, 87, 29}
465 {1, 3, 5, 15, 465, 155, 93, 31}
496 {1, 2, 4, 8, 16, 496, 248, 124, 62, 

76576500

In [69]:
# first ten digits of sum of the below

digits = r"""
37107287533902102798797998220837590246510135740250
46376937677490009712648124896970078050417018260538
74324986199524741059474233309513058123726617309629
91942213363574161572522430563301811072406154908250
23067588207539346171171980310421047513778063246676
89261670696623633820136378418383684178734361726757
28112879812849979408065481931592621691275889832738
44274228917432520321923589422876796487670272189318
47451445736001306439091167216856844588711603153276
70386486105843025439939619828917593665686757934951
62176457141856560629502157223196586755079324193331
64906352462741904929101432445813822663347944758178
92575867718337217661963751590579239728245598838407
58203565325359399008402633568948830189458628227828
80181199384826282014278194139940567587151170094390
35398664372827112653829987240784473053190104293586
86515506006295864861532075273371959191420517255829
71693888707715466499115593487603532921714970056938
54370070576826684624621495650076471787294438377604
53282654108756828443191190634694037855217779295145
36123272525000296071075082563815656710885258350721
45876576172410976447339110607218265236877223636045
17423706905851860660448207621209813287860733969412
81142660418086830619328460811191061556940512689692
51934325451728388641918047049293215058642563049483
62467221648435076201727918039944693004732956340691
15732444386908125794514089057706229429197107928209
55037687525678773091862540744969844508330393682126
18336384825330154686196124348767681297534375946515
80386287592878490201521685554828717201219257766954
78182833757993103614740356856449095527097864797581
16726320100436897842553539920931837441497806860984
48403098129077791799088218795327364475675590848030
87086987551392711854517078544161852424320693150332
59959406895756536782107074926966537676326235447210
69793950679652694742597709739166693763042633987085
41052684708299085211399427365734116182760315001271
65378607361501080857009149939512557028198746004375
35829035317434717326932123578154982629742552737307
94953759765105305946966067683156574377167401875275
88902802571733229619176668713819931811048770190271
25267680276078003013678680992525463401061632866526
36270218540497705585629946580636237993140746255962
24074486908231174977792365466257246923322810917141
91430288197103288597806669760892938638285025333403
34413065578016127815921815005561868836468420090470
23053081172816430487623791969842487255036638784583
11487696932154902810424020138335124462181441773470
63783299490636259666498587618221225225512486764533
67720186971698544312419572409913959008952310058822
95548255300263520781532296796249481641953868218774
76085327132285723110424803456124867697064507995236
37774242535411291684276865538926205024910326572967
23701913275725675285653248258265463092207058596522
29798860272258331913126375147341994889534765745501
18495701454879288984856827726077713721403798879715
38298203783031473527721580348144513491373226651381
34829543829199918180278916522431027392251122869539
40957953066405232632538044100059654939159879593635
29746152185502371307642255121183693803580388584903
41698116222072977186158236678424689157993532961922
62467957194401269043877107275048102390895523597457
23189706772547915061505504953922979530901129967519
86188088225875314529584099251203829009407770775672
11306739708304724483816533873502340845647058077308
82959174767140363198008187129011875491310547126581
97623331044818386269515456334926366572897563400500
42846280183517070527831839425882145521227251250327
55121603546981200581762165212827652751691296897789
32238195734329339946437501907836945765883352399886
75506164965184775180738168837861091527357929701337
62177842752192623401942399639168044983993173312731
32924185707147349566916674687634660915035914677504
99518671430235219628894890102423325116913619626622
73267460800591547471830798392868535206946944540724
76841822524674417161514036427982273348055556214818
97142617910342598647204516893989422179826088076852
87783646182799346313767754307809363333018982642090
10848802521674670883215120185883543223812876952786
71329612474782464538636993009049310363619763878039
62184073572399794223406235393808339651327408011116
66627891981488087797941876876144230030984490851411
60661826293682836764744779239180335110989069790714
85786944089552990653640447425576083659976645795096
66024396409905389607120198219976047599490197230297
64913982680032973156037120041377903785566085089252
16730939319872750275468906903707539413042652315011
94809377245048795150954100921645863754710598436791
78639167021187492431995700641917969777599028300699
15368713711936614952811305876380278410754449733078
40789923115535562561142322423255033685442488917353
44889911501440648020369068063960672322193204149535
41503128880339536053299340368006977710650566631954
81234880673210146739058568557934581403627822703280
82616570773948327592232845941706525094512325230608
22918802058777319719839450180888072429661980811197
77158542502016545090413245809786882778948721859617
72107838435069186155435662884062257473692284509516
20849603980134001723930671666823555245252804609722
53503534226472524250874054075591789781264330331690
"""

digits = digits.strip()

digits = digits.split("\n")

In [70]:

total = []
overflow = 0
linelen = len(digits[0])
# count backwards, assume all have same 

for i in range(1, linelen+1):
    current = 0
    idx = -i
    for line in digits:
        current += int(line[idx])
    current += overflow
    current = str(current)
    total.insert(0, current[-1])
    if len(current) > 1:
        overflow = int(current[:-1])
    else:
        overflow = 0

if overflow:
    total.insert(0, str(overflow))

total = ''.join(total)

In [71]:
total

'5537376230390876637302048746832985971773659831892672'

In [72]:
len(total)

52

The following iterative sequence is defined for the set of positive integers:

n → n/2 (n is even)
n → 3n + 1 (n is odd)

Using the rule above and starting with 13, we generate the following sequence:

13 → 40 → 20 → 10 → 5 → 16 → 8 → 4 → 2 → 1
It can be seen that this sequence (starting at 13 and finishing at 1) contains 10 terms. Although it has not been proved yet (Collatz Problem), it is thought that all starting numbers finish at 1.

Which starting number, under one million, produces the longest chain?

NOTE: Once the chain starts the terms are allowed to go above one million.



In [97]:
last = 1000000
# keys: start digit, values: path length
lengths = {1: 0}
maxn = 1

for n in range(2, last+1):
    current = n
    steps = 0
    while True:
        # if it reduces to a previous n, add the current steps
        # and stop
        if current in lengths:
            lengths[n] = steps + lengths[current]
            break
        if current == 1:
            lengths[n] = steps
            break
        elif current % 2 == 0:
            current = current/2
        else:
            current = current*3 + 1
        steps += 1
    if lengths[n] > lengths[maxn]:
        maxn = n
        

In [98]:
maxn, lengths[maxn]

(837799, 524)

In [99]:
# Number of routes from top left corner to bottom right corner of 20x20 grid
# --> 20 choose 10

def fact(n):
    return reduce(mul, range(1, n+1))

In [101]:
f20 = fact(20)
f10 = fact(10)

f20/(f10**2)

184756.0

$2^{15} = 32768$ and the sum of its digits is 3 + 2 + 7 + 6 + 8 = 26.

What is the sum of the digits of the number $2^{1000}$?



In [105]:
for n in range(1, 20):
    s = 2**n
    sumdigits = sum(map(int, str(s)))
    print('{}: {}'.format(s, sumdigits))
    

2: 2
4: 4
8: 8
16: 7
32: 5
64: 10
128: 11
256: 13
512: 8
1024: 7
2048: 14
4096: 19
8192: 20
16384: 22
32768: 26
65536: 25
131072: 14
262144: 19
524288: 29


In [106]:
def add_strings(digits):
    # digits: list of strings describing the digits
    # value is returned as a string. 
    total = []
    overflow = 0
    linelen = len(digits[0])
    # count backwards, assume all have same 

    for i in range(1, linelen+1):
        current = 0
        idx = -i
        for line in digits:
            current += int(line[idx])
        current += overflow
        current = str(current)
        total.insert(0, current[-1])
        if len(current) > 1:
            overflow = int(current[:-1])
        else:
            overflow = 0

    if overflow:
        total.insert(0, str(overflow))

    total = ''.join(total)
    return total

def powerof2(n):
    if n == 1:
        return '1'
    # do this using the addition routine with strings developed earlier. 
    digits = ['1', '1']
    for i in range(1, n+1):
        out = add_strings(digits)
        digits = [out, out]
    return out


In [111]:
str2_1000 = powerof2(1000)

In [112]:
sum(map(int, str2_1000))

1366

If the numbers 1 to 5 are written out in words: one, two, three, four, five, then there are 3 + 3 + 5 + 4 + 4 = 19 letters used in total.

If all the numbers from 1 to 1000 (one thousand) inclusive were written out in words, how many letters would be used?


NOTE: Do not count spaces or hyphens. For example, 342 (three hundred and forty-two) contains 23 letters and 115 (one hundred and fifteen) contains 20 letters. The use of "and" when writing out numbers is in compliance with British usage.

In [139]:
dictionary = {
    1: 'one',
    2: 'two',
    3: 'three',
    4: 'four',
    5: 'five',
    6: 'six',
    7: 'seven',
    8: 'eight',
    9: 'nine',
    10: 'ten',
    11: 'eleven',
    12: 'twelve',
    13: 'thirteen',
    14: 'fourteen',
    15: 'fifteen',
    16: 'sixteen',
    17: 'seventeen',
    18: 'eighteen',
    19: 'nineteen',
    20: 'twenty',
    30: 'thirty',
    40: 'forty',
    50: 'fifty',
    60: 'sixty',
    70: 'seventy',
    80: 'eighty',
    90: 'ninety',
    100: 'hundred',
    1000: 'thousand',
}

def decompose(n):
    return {
        1000: n//1000,
        100: (n%1000)//100,
        10: (n%100)//10,
        1: n%10,
    }

def format(n):
    factors = decompose(n)
    #print(factors)
    retval = []
    ones = factors[1]
    tens = factors[10]
    hundreds = factors[100]
    thousands = factors[1000]
    # combination of ones and tens determines whether we will use words
    ones_tens = ones + tens*10
    if ones_tens and ones_tens <= 20:
        retval.append(dictionary[ones_tens])
    else:
        if ones:
            if not tens:
                retval.append(dictionary[ones])
            else:
                multiple = tens*10
                retval.append('{}-{}'.format(dictionary[multiple], dictionary[ones]))
        else:
            if tens:
                multiple = tens*10
                retval.append(dictionary[multiple])
    if hundreds:
        # we only want to add "and" if there is already something. 
        if retval:
            retval.insert(0, 'and')
        retval.insert(0, '{} hundred'.format(dictionary[hundreds]))
    if thousands:
        if retval:
            retval.insert(0, 'and')
        retval.insert(0, '{} thousand'.format(dictionary[thousands]))
    return " ".join(retval)

In [140]:
inwords = [format(i) for i in range(1, 1001)]

numletters = len(''.join(inwords).replace(' ', '').replace('-', ''))

In [141]:
numletters

21124

By starting at the top of the triangle below and moving to adjacent numbers on the row below, the maximum total from top to bottom is 23.

3

7 4

2 4 6

8 5 9 3

That is, 3 + 7 + 4 + 9 = 23.

Find the maximum total from top to bottom of the triangle below:

75

95 64

17 47 82

18 35 87 10

20 04 82 47 65

19 01 23 75 03 34

88 02 77 73 07 63 67

99 65 04 28 06 16 70 92

41 41 26 56 83 40 80 70 33

41 48 72 33 47 32 37 16 94 29

53 71 44 65 25 43 91 52 97 51 14

70 11 33 28 77 73 17 78 39 68 17 57

91 71 52 38 17 14 91 43 58 50 27 29 48

63 66 04 68 89 53 67 30 73 16 69 87 40 31

04 62 98 27 23 09 70 98 73 93 38 53 60 04 23

NOTE: As there are only 16384 routes, it is possible to solve this problem by trying every route. However, Problem 67, is the same challenge with a triangle containing one-hundred rows; it cannot be solved by brute force, and requires a clever method! ;o)

In [157]:
import copy
triangle = r"""
75
95 64
17 47 82
18 35 87 10
20 04 82 47 65
19 01 23 75 03 34
88 02 77 73 07 63 67
99 65 04 28 06 16 70 92
41 41 26 56 83 40 80 70 33
41 48 72 33 47 32 37 16 94 29
53 71 44 65 25 43 91 52 97 51 14
70 11 33 28 77 73 17 78 39 68 17 57
91 71 52 38 17 14 91 43 58 50 27 29 48
63 66 04 68 89 53 67 30 73 16 69 87 40 31
04 62 98 27 23 09 70 98 73 93 38 53 60 04 23
"""

triangle = triangle.strip()
triangle = triangle.split("\n")
triangle = [line.split( ) for line in triangle]

def maxpath(triangle):
    # try from bottom up to reduce the triangle.
    # Start with the second last row, calculate as if
    # we had chosen the largest value in the last row.    
    # for the moment, make a copy of the triangle
    # so we can use a destructive algorithm
    t2 = copy.deepcopy(triangle)
    depth = len(triangle)
    for thisidx, nextidx in zip(range(depth-2, -1, -1), range(depth-1, 0, -1)):
        thisrow = t2[thisidx]
        nextrow = t2[nextidx]
        for jdx, item in enumerate(map(int, thisrow)):
            left, right = nextrow[jdx:jdx+2]
            t2[thisidx][jdx] = item + max(int(left), int(right))
        #print(thisidx, t2)
    return t2

maxpath(triangle)
        
    

[[1074],
 [995, 999],
 [818, 900, 935],
 [704, 801, 853, 792],
 [686, 640, 766, 731, 782],
 [666, 614, 636, 684, 660, 717],
 [647, 501, 613, 609, 533, 657, 683],
 [559, 499, 479, 536, 514, 526, 594, 616],
 [460, 434, 419, 475, 508, 470, 510, 524, 487],
 [419, 365, 393, 387, 419, 425, 430, 376, 454, 322],
 [378, 317, 231, 321, 354, 372, 393, 354, 360, 293, 247],
 [325, 246, 187, 178, 256, 329, 273, 302, 263, 242, 193, 233],
 [255, 235, 154, 150, 140, 179, 256, 209, 224, 172, 174, 176, 148],
 [125, 164, 102, 95, 112, 123, 165, 128, 166, 109, 122, 147, 100, 54],
 ['04',
  '62',
  '98',
  '27',
  '23',
  '09',
  '70',
  '98',
  '73',
  '93',
  '38',
  '53',
  '60',
  '04',
  '23']]

You are given the following information, but you may prefer to do some research for yourself.

* 1 Jan 1900 was a Monday.
* Thirty days has September, April, June and November.
* All the rest have thirty-one,
* Saving February alone, Which has twenty-eight, rain or shine.
* And on leap years, twenty-nine.
* A leap year occurs on any year evenly divisible by 4, but not on a century unless it is divisible by 400.
* How many Sundays fell on the first of the month during the twentieth century (1 Jan 1901 to 31 Dec 2000)?



In [179]:
# Let Monday be encoded as 0 because it's easy
# algorithm:
# assign each day of each year a cardinal number
# for each year, calculate the day of week of every first day of month
# keep track of the Sundays. 

def is_leap_year(n):
    if n%100 == 0:
        if n%400 == 0:
            return True
        else:
            return False
    elif n%4 == 0:
        return True
    return False

def days_in_month(y, m):
    if m == 2:
        if is_leap_year(y):
            return 29
        else:
            return 28
    elif m in (1, 3, 5, 7, 8, 10, 12):
          return 31
    return 30
    
def weekday_firstdayof_month(y, m):
    # number of years between n and 1900
    between = y-1900
    days = sum([(366 if is_leap_year(i) else 365) for i in range(1900, y)])
    # add on the months
    days += sum([days_in_month(y, i) for i in range(1, m)])
    # first day of 1900 was Monday which is weekday 0
    mod = days%7
    # "mod" gives the number of weekdays offset from Monday
    return mod

collect = {}
for year in range(1900, 2001):
    for month in range(1, 13):
        wd = weekday_firstdayof_month(year, month)
        if wd == 6:
            collect[(year, month)] = True

In [180]:
collect

{(1900, 4): True,
 (1900, 7): True,
 (1901, 9): True,
 (1901, 12): True,
 (1902, 6): True,
 (1903, 2): True,
 (1903, 3): True,
 (1903, 11): True,
 (1904, 5): True,
 (1905, 1): True,
 (1905, 10): True,
 (1906, 4): True,
 (1906, 7): True,
 (1907, 9): True,
 (1907, 12): True,
 (1908, 3): True,
 (1908, 11): True,
 (1909, 8): True,
 (1910, 5): True,
 (1911, 1): True,
 (1911, 10): True,
 (1912, 9): True,
 (1912, 12): True,
 (1913, 6): True,
 (1914, 2): True,
 (1914, 3): True,
 (1914, 11): True,
 (1915, 8): True,
 (1916, 10): True,
 (1917, 4): True,
 (1917, 7): True,
 (1918, 9): True,
 (1918, 12): True,
 (1919, 6): True,
 (1920, 2): True,
 (1920, 8): True,
 (1921, 5): True,
 (1922, 1): True,
 (1922, 10): True,
 (1923, 4): True,
 (1923, 7): True,
 (1924, 6): True,
 (1925, 2): True,
 (1925, 3): True,
 (1925, 11): True,
 (1926, 8): True,
 (1927, 5): True,
 (1928, 1): True,
 (1928, 4): True,
 (1928, 7): True,
 (1929, 9): True,
 (1929, 12): True,
 (1930, 6): True,
 (1931, 2): True,
 (1931, 3): Tru

In [181]:
len(collect)

173

n! means n × (n − 1) × ... × 3 × 2 × 1

For example, 10! = 10 × 9 × ... × 3 × 2 × 1 = 3628800,

and the sum of the digits in the number 10! is 3 + 6 + 2 + 8 + 8 + 0 + 0 = 27.

Find the sum of the digits in the number 100!



In [189]:
def sumofdigits_factorial(n):
    # find the sum of digits in n!
    # throw away any trailing zeros and keep multiplying.
    total = 1
    # current sum of digits
    current = 0
    i = 0
    while i < n:
        i += 1

        total *= i
        if total%10 == 0:
            while total %10 == 0:
                total //=10
        current = sum(map(int, str(total)))
    return total, current
    

In [190]:
sumofdigits_factorial(100)

(93326215443944152681699238856266700490715968264381621468592963895217599993229915608941463976156518286253697920827223758251185210916864,
 648)